// @sveltejs/kit/dist/index4.js
'use strict';

var index = require('./index.js');
var path = require('path');
require('module');
var Url = require('url');
var fs = require('fs');
var index$1 = require('./index5.js');
require('./standard.js');
var http = require('http');

const mutable = (dir) =>
	index$1.sirv(dir, {
		etag: true,
		maxAge: 0
	});

function start({ port, config }) {
	return new Promise((fulfil) => {
		const app = index.requireRelative_1('./.svelte/build/optimized/server/app.js');

		const static_handler = fs.existsSync(config.files.assets)
			? mutable(config.files.assets)
			: (_req, _res, next) => next();

		const assets_handler = index$1.sirv('.svelte/build/optimized/client', {
			maxAge: 31536000,
			immutable: true
		});

		const server = http.createServer((req, res) => {
			const parsed = Url.parse(req.url || '');

			assets_handler(req, res, () => {
				static_handler(req, res, async () => {
					const rendered = await app.render(
						{
							method: req.method,
							headers: req.headers,
							path: parsed.pathname,
							body: await index$1.get_body(req),
							query: new Url.URLSearchParams(parsed.query || '')
						},
						{
							paths: {
								base: '',
								assets: '/.'
							},
							get_stack: (error) => error.stack, // TODO should this return a sourcemapped stacktrace?
							get_static_file: (file) => fs.readFileSync(path.join(config.files.assets, file))
						}
					);

					if (rendered) {
						res.writeHead(rendered.status, rendered.headers);
						res.end(rendered.body);
					} else {
						res.statusCode = 404;
						res.end('Not found');
					}
				});
			});
		});

		server.listen(port, () => {
			fulfil(server);
		});

		return server;
	});
}

exports.start = start;
